lib/oli_web/live/system_message_live/index_view.ex

```diff
@@ -0,0 +1,255 @@
+      [
+        Breadcrumb.new(%{
+          full_title: "New System Message",
+          link: Routes.live_path(OliWeb.Endpoint, __MODULE__)
+        })
+      ]
+  end
+
+  def mount(_, %{"local_tz" => local_tz}, socket) do
+    messages = Notifications.list_system_messages()
+
+    {:ok,
+     assign(socket,
+       messages: messages,
+       local_tz: local_tz,
+       breadcrumbs: breadcrumb()
+     )}
+  end
+
+  def render(assigns) do
+    ~F"""
+      {#for message <- @messages}
+        <EditMessage system_message={current_message(@unsaved_system_message, message)} timezone={@local_tz}/>
+      {/for}
+      <Form for={:system_message} submit="create">
+        <Field name={:message} class="form-group">
+          <TextArea
+            class="form-control"
+            rows="3"
+            opts={placeholder: "Type a message for all users in the system", maxlength: "140"}
+          />
+          <ErrorTag/>
+        </Field>
+
+        <button class="form-button btn btn-md btn-primary btn-block mt-3" type="submit">Create</button>
+      </Form>
+      {#if @show_confirm}
+        <Confirm title="Confirm Message" id="dialog" ok="broadcast_message" cancel="cancel_modal">
+          Are you sure that you wish to <b>{if @message_will_be_displayed, do: "send", else: "hide"}</b> this message to all users in the system?
+        </Confirm>
+      {/if}
+    """
+  end
+
+  def handle_event("create", %{"system_message" => %{"message" => message}}, socket) do
+    socket = clear_flash(socket)
+
+    case Notifications.create_system_message(%{message: message}) do
+      {:ok, _system_message} ->
+        {:noreply,
+         socket
+         |> put_flash(:info, "System message successfully created.")
+         |> assign(messages: Notifications.list_system_messages())}
+
+      {:error, %Ecto.Changeset{} = changeset} ->
+        socket =
+          put_flash(
+            socket,
+            :error,
+            "System message couldn't be created: #{translate_all_changeset_errors(changeset)}."
+          )
+
+        {:noreply, socket}
+    end
+  end
+
+  def handle_event(
+        "save",
+        %{"system_message" => %{"active" => active, "id" => id} = attrs},
+        socket
+      ) do
+    socket = clear_flash(socket)
+
+    active = Oli.Utils.string_to_boolean(active)
+    id = String.to_integer(id)
+    system_message = find_system_message(id, socket.assigns.messages)
+
+    message_attrs =
+      attrs
+      |> Oli.Utils.atomize_keys()
+      |> Map.update(:start, "", &datestring_to_utc_datetime(&1, socket.assigns.local_tz))
+      |> Map.update(:end, "", &datestring_to_utc_datetime(&1, socket.assigns.local_tz))
+      |> Map.put(:active, active)
+      |> Map.put(:id, id)
+
+    message_was_being_displayed =
+      system_message
+      |> Map.from_struct()
+      |> message_displayed?()
+
+    message_will_be_displayed = message_displayed?(message_attrs)
+
+    # only show confirmation modal when:
```

Comment:
I'm glad you added the comments, I had trouble quickly understanding what was going on here - wonder if it's worth moving into a private function where it could be named well, and the complexity moved there?